README.md
# amartus
    Initial README.md for amartus

# how to run it

## how to run a server:
    export FLASK_APP=amartus_app.py
    flask run

## how to run a client:
### to run help:
    python3 amartus/client.py --help

### to run the app:
    python3 amartus/client.py http://my.server.address

CHANGELOG.md
# Changelog

## [0.0.1] - 2019-09-16
### Added
- Initial commit

## [0.1.0] - 2019-09-18
### Initial version

tests/test_client.py
import json
from unittest import TestCase
from unittest.mock import patch, MagicMock
import amartus.client as client
from amartus.client import getnode, get_ip


class TestClient(TestCase):
    def setUp(self):
        self.patcher = patch('amartus.client.urlopen')
        self.mocked_urlopen = self.patcher.start()

    def tearDown(self):
        self.patcher.stop()

    def test_args(self):
        args = client.parse_args([
            'https:/my.local.address',
            '-n',
            'Peter',
            '-u',
            'Data'])
        self.assertEqual(args.name, 'Peter')
        self.assertEqual(args.userdata, 'Data')

    def test_get_ip(self):
        # Don't really have an idea how to test this functionality
        pass

    def test_main(self):
        with patch('amartus.client.parse_args') as mocked_parse:
            with patch('builtins.print'):
                local_args = MagicMock()
                local_args.name = 'Peter'
                local_args.userdata = 'Data'
                local_args.server_address = 'http://my.local.address'
                mocked_parse.return_value = local_args
                self.mocked_urlopen.return_value.read\
                    .return_value.decode.return_value\
                    = '{"register": "SUCCESS"}'
                client.main()

        args, kwargs = self.mocked_urlopen.call_args
        request = args[0]
        self.assertEqual(
            request.full_url,
            'http://my.local.address/ztp/register/{}'.format(getnode()))
        self.assertEqual(
            json.loads(request.data),
            {'ip': get_ip(), 'name': 'Peter', 'userdata': 'Data'}
        )
        self.assertEqual(
            request.header_items(),
            [('Content-type', 'application/json')]
        )


tests/test_routes.py
import json
from urllib.error import URLError
from unittest import TestCase
from unittest.mock import patch
from amartus import main_app, main_app_register


class TestRoutes(TestCase):
    def setUp(self):
        main_app.config['TESTING'] = True
        self.app = main_app.test_client()
        main_app_register._register = {}
        self.patcher = patch('amartus.routes.urlopen')
        self.mocked_urlopen = self.patcher.start()

    def tearDown(self):
        self.patcher.stop()

    def test_main_route(self):
        result = self.app.get('/')
        self.assertEqual(result.status_code, 404)

    def test_register_route(self):
        with patch('amartus.routes.main_app_register') as mocked_register:
            response = self.app.put(
                '/ztp/register/1',
                data=json.dumps({
                    'ip': '1.1.1.1',
                    'name': 'Peter',
                    'userdata': 'My special string'}),
                follow_redirects=True)
        calls = []
        for call in mocked_register.mock_calls:
            name, args, kwargs = call
            calls.append((name, args, kwargs))
        self.assertEqual(
            calls,
            [('register', ('1', '1.1.1.1', 'Peter', 'My special string'), {})])
        calls = []
        for call in self.mocked_urlopen.call_args_list:
            args, kwargs = call
            calls.append((args, kwargs))
        self.assertEqual(
            calls,
            [(('http://adress/api/exec?ip=1.1.1.1',), {})]
        )
        self.assertEqual(
            json.loads(response.data.decode()),
            {'register': 'SUCCESS'})
        response = self.app.put(
            '/ztp/register/1',
            data={
                'ip': '1.11.1.1'
            }
        )
        self.assertEqual(response.status_code, 400)
        self.app.put(
            '/ztp/register/1',
            data=json.dumps({
                'ip': '1.1.1.1',
                'name': 'Peter',
                'userdata': 'My special string'}),
            follow_redirects=True)
        response = self.app.put(
            '/ztp/register/1',
            data=json.dumps({
                'ip': '1.1.1.1',
                'name': 'Peter',
                'userdata': 'My special string'}),
            follow_redirects=True)
        self.assertEqual(response.status_code, 409)

        def raiseError(*args, **kwargs):
            raise URLError('error')

        self.mocked_urlopen.side_effect = raiseError
        response = self.app.put(
            '/ztp/register/5',
            data=json.dumps({
                'ip': '1.1.1.1',
                'name': 'Peter',
                'userdata': 'My special string'}),
            follow_redirects=True)
        self.assertEqual(
            json.loads(response.data.decode()),
            {'register': 'FAILED TO EXECUTE'})

    def test_check_route(self):
        self.app.put(
            '/ztp/register/1',
            data=json.dumps({
                'ip': '1.1.1.1',
                'name': 'Peter',
                'userdata': 'MyData'
            }),
            follow_redirects=True
        )
        response = self.app.get('/ztp/check/1')
        self.assertEqual(
            {'ip': '1.1.1.1', 'name': 'Peter', 'userdata': 'MyData'},
            json.loads(response.data.decode()))
        response = self.app.get('/ztp/check/2')
        self.assertEqual({}, json.loads(response.data.decode()))

    def test_list_route(self):
        self.app.put(
            '/ztp/register/1',
            data=json.dumps({
                'ip': '1.1.1.1',
                'name': 'Peter',
                'userdata': 'MyData'
            }),
            follow_redirects=True)
        self.app.put(
            '/ztp/register/2',
            data=json.dumps({
                'ip': '1.1.3.1',
                'name': 'Ela',
                'userdata': 'ElaData'
            }),
            follow_redirects=True)
        response = self.app.get('/ztp/mng/list')
        self.assertEqual(
            {'hosts': ['1', '2']},
            json.loads(response.data.decode()))

    def test_delete_route(self):
        self.app.put(
            '/ztp/register/1',
            data=json.dumps({
                'ip': '1.1.1.1',
                'name': 'Peter',
                'userdata': 'MyData'
            }),
            follow_redirects=True)
        response = self.app.get('ztp/check/1')
        self.assertTrue(json.loads(response.data.decode()))
        self.app.get('/ztp/mng/1/delete')
        response = self.app.get('/ztp/check/1')
        self.assertFalse(json.loads(response.data.decode()))

    def test_edit_route(self):
        self.app.put(
            '/ztp/register/1',
            data=json.dumps({
                'ip': '1.1.1.1',
                'name': 'Peter',
                'userdata': 'MyData'
            }),
            follow_redirects=True)
        self.app.put(
            '/ztp/mng/1',
            data=json.dumps({
                'ip': '2.2.2.2'
            }),
            follow_redirects=True
        )
        response = self.app.get('/ztp/check/1')
        self.assertEqual(
            json.loads(response.data.decode()),
            {
                'ip': '2.2.2.2',
                'name': 'Peter',
                'userdata': 'MyData'
            }
        )
        response = self.app.put(
            '/ztp/mng/1',
            data=json.dumps({
                'sss': 'error'
            }),
            follow_redirects=True
        )
        self.assertEqual(response.status_code, 400)
        response = self.app.put(
            '/ztp/mng/3',
            data=json.dumps({
                'ip': '1.1.1.1'
            })
        )
        self.assertEqual(response.status_code, 400)


tests/test_amartus.py
from amartus import main_app
from unittest import TestCase


class FirstTestCase(TestCase):
    def setUp(self):
        main_app.config['TESTING'] = True
        main_app.config['WTF_CSRF_ENABLED'] = False
        main_app.config['DEBUG'] = False
        self.app = main_app.test_client()

    def tearDown(self):
        pass

    def test_main_route(self):
        result = self.app.get('/')
        self.assertEqual(result.status_code, 404)

    def test_register_route(self):
        result = self.app.put(
            '/ztp/register/1',
            data={
                'ip': '1.1.1.1',
                'name': 'Peter',
                'userdata': 'My special string'
            },
            follow_redirects=True)
        print(result.data)


amartus/client.py
import sys
import argparse
import socket
import json
from uuid import getnode
from urllib.request import urlopen, Request
from urllib.parse import urljoin
from urllib.error import HTTPError


def parse_args(args=sys.argv[1:]):
    parser = argparse.ArgumentParser()
    parser.add_argument(
        'server_address',
        help='Server address'
    )
    parser.add_argument(
        '-n',
        '--name',
        help='Name',
        default=''
    )
    parser.add_argument(
        '-u',
        '--userdata',
        help='Userdata',
        default=''
    )
    return parser.parse_args(args)


def get_ip():
    s = socket.socket(socket.AF_INET, socket.SOCK_DGRAM)
    try:
        s.connect(('10.255.255.255', 1))
        ip = s.getsockname()[0]
    except Exception:
        ip = '127.0.0.1'
    finally:
        s.close()
    return ip


def main():
    args = parse_args()
    new_register_request = Request(
        urljoin(args.server_address, '/ztp/register/{}'.format(getnode())),
        data=json.dumps({
            "ip": get_ip(),
            "name": args.name,
            "userdata": args.userdata
        }).encode(),
        method='PUT',
        headers={'Content-type': 'application/json'}
    )
    try:
        response = urlopen(new_register_request)
    except HTTPError:
        print('register conflict')
        return 1
    print(json.loads(response.read().decode()))
    return 0


if __name__ == '__main__':
    exit(main())


amartus/register.py
class Register:
    def __init__(self):
        self._register = {}

    def register(self, id, ip, name, userdata):
        if id in self._register:
            raise ValueError('{} already registered'.format(id))
        self._register[id] = {'ip': ip, 'name': name, 'userdata': userdata}

    def check(self, id):
        if id in self._register:
            return self._register[id]
        return {}

    def list_register(self):
        return sorted(self._register.keys())

    def delete(self, id):
        del self._register[id]

    def update(self, id, ip=None, name=None, userdata=None):
        if id not in self._register:
            raise KeyError(id)
        if ip is not None:
            self._register[id]['ip'] = ip
        if name is not None:
            self._register[id]['name'] = name
        if userdata is not None:
            self._register[id]['userdata'] = userdata


amartus/routes.py
import json
from urllib.request import urlopen
from urllib.error import URLError
from flask import jsonify, abort
from amartus import main_app, main_app_register, request


@main_app.route('/ztp/register/<id>', methods=['PUT'])
def register(id):
    try:
        data = json.loads(request.data.decode())
    except Exception:
        abort(400)
    try:
        main_app_register.register(
            id,
            data['ip'],
            data['name'],
            data['userdata'])
    except ValueError:
        abort(409)
    try:
        urlopen('http://adress/api/exec?ip={}'.format(data['ip']))
    except URLError:
        return jsonify({'register': "FAILED TO EXECUTE"})
    return jsonify({'register': 'SUCCESS'})


@main_app.route('/ztp/check/<id>', methods=['GET'])
def check(id):
    return jsonify(main_app_register.check(id))


@main_app.route('/ztp/mng/list', methods=['GET'])
def list_register():
    register_list = main_app_register.list_register()
    return jsonify({'hosts': register_list})


@main_app.route('/ztp/mng/<id>/delete', methods=['GET'])
def delete(id):
    main_app_register.delete(id)
    return jsonify({'delete': 'SUCCESS'})


@main_app.route('/ztp/mng/<id>', methods=['PUT'])
def edit(id):
    try:
        data = json.loads(request.data.decode())
    except Exception:
        abort(400)
    good_keys = {'ip', 'name', 'userdata'}
    if not good_keys >= set(data.keys()):
        abort(400)
    try:
        main_app_register.update(id, **data)
    except KeyError:
        abort(400)
    return jsonify(main_app_register.check(id))
